Document/0x05e-Testing-Cryptography.md

```diff
@@ -8,21 +8,76 @@ A general rule in app development is that one should never attempt to invent the
 
 Android developers don't need to bother much with the intricate details of cryptography most of the time. However, even when using standard algorithms can be affected if misconfigured. 
 
-#### Static Analysis
+Android SDK provides mechanisms for specifying secure key generation and use. Android 6.0 (Marshmallow, API 23) introduced the `KeyGenParameterSpec` class that can be used to ensure the correct key usage in the application.
+
+Here's an example of using AES:
 
--- TODO [Describe Static Analysis on Verifying the Configuration of Cryptographic Standard Algorithms : how to assess this given either the source code or installer package (APK/IPA/etc.), but without running the app. Tailor this to the general situation (e.g., in some situations, having the decompiled classes is just as good as having the original source, in others it might make a bigger difference). If required, include a subsection about how to test with or without the original sources.] --
+```
+String keyAlias = "MySecretKey";
+String ANDROID_KEY_STORE = "AndroidKeyStore";
 
--- TODO [Clarify the purpose of "Use the &lt;sup&gt; tag to reference external sources, e.g. Meyer's recipe for tomato soup<sup>[1]</sup>."] --
+KeyGenParameterSpec keyGenParameterSpec = new KeyGenParameterSpec.Builder(keyAlias,
+        KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+        .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+        .setRandomizedEncryptionRequired(true)
+        .build();
 
--- TODO [Develop Static Analysis with source code of "Verifying the Configuration of Cryptographic Standard Algorithms"] --
+KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES,
+        ANDROID_KEY_STORE);
+keyGenerator.init(keyGenParameterSpec);
 
-#### Dynamic Analysis
+SecretKey secretKey = keyGenerator.generateKey();
+```
+
+The `KeyGenParameterSpec` indicates that the key can be used for encryption and decryption, but not for other purposes, such as signing or verifying. It further specifies the block mode (GCM), padding (none), and explicitly specifies that randomized encryption is required (this is the default.)
+
+Attempting to use the generated key in violation of the above spec would result in a security exception.
+
+Here's an example of using that key correctly:
+
+```
+String AES_MODE = KeyProperties.KEY_ALGORITHM_AES
+        + "/" + KeyProperties.BLOCK_MODE_GCM
+        + "/" + KeyProperties.ENCRYPTION_PADDING_NONE;
+KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
+
+Key key = keyStore.getKey(keyAlias, null);
+
+Cipher cipher = Cipher.getInstance(AES_MODE);
+cipher.init(Cipher.ENCRYPT_MODE, key);
+
+byte[] encryptedBytes = cipher.doFinal(input);
+byte[] iv = cipher.getIV();
+```
+
```

Comment:
Maybe add a reference to the google developer documentation on the asymmetric keypairgenerationspec?